checkbox-ng/checkbox_ng/config.py

```diff
@@ -33,51 +33,98 @@
+    return to_r
+
+
 def load_configs(launcher_file=None, cfg=None):
     """
     Read a chain of configs/launchers.
 
     In theory there can be a very long list of configs that are linked by
-    specifying config_filename in each. Each time this happen we need to
-    consider the new one and override all the values contained therein.
-    And after this chain is exhausted the values in the launcher should
-    take precedence over the previously read.
-    Warning: some recursion ahead.
+    specifying config_filename in each. Each config that defines a
+    config_filename imports the values that are defined in the path/name
+    provided. The imported are overwritten by the importee and whoever has
+    an higher priority.
+
+    Ex: If ~/.config/checkbox.conf has config_filename A and we have both
+        ~/.config/A and /etc/xdg/A:
+
+        - ~/.config/A has an higher priority than /etc/xdg/A
+        - anything that ~/.config/A imports has a higher priority
+            than /etc/xdg/A but lower than ~/.config/A itself
+        - anything that /etc/xdg/A imports has the lowest possible
+            priority
     """
+    assert not (launcher_file and cfg), "config_filename in cfg will be ignored, FIXME"
     if not cfg:
         cfg = Configuration()
-    previous_cfg_name = cfg.get_value('config', 'config_filename')
-    if os.path.isabs(expand_all(previous_cfg_name)):
-        cfg.update_from_another(
-            Configuration.from_path(expand_all(previous_cfg_name)),
-            'config file: {}'.format(previous_cfg_name))
+    if launcher_file:
+        # Use the config_filename if it is defined in launcher
+        launcher_file_conf = Configuration.from_path(launcher_file)
+        to_load_conf_names = _search_configs_by_name(
+            launcher_file_conf.get_value("config", "config_filename")
+        )
     else:
-        for sdir in SEARCH_DIRS:
-            config = expand_all(os.path.join(sdir, previous_cfg_name))
-            if os.path.exists(config):
-                cfg.update_from_another(
-                    Configuration.from_path(config),
-                    'config file: {}'.format(config))
-            else:
-                _logger.info(
-                    "Referenced config file doesn't exist: %s", config)
-    new_cfg_filename = cfg.get_value('config', 'config_filename')
-    if new_cfg_filename != previous_cfg_name:
-        load_configs(launcher_file, cfg)
+        # configs to read who may have a redirect in them
+        to_load_conf_names = _search_configs_by_name(
+            cfg.get_value("config", "config_filename")
+        )
+    # used to avoid "loops"
+    # Note: checkbox.conf is always the default "config_filename"
+    #       so we *always* have a loop eventually
+    already_loaded = {cfg.get_value("config", "config_filename")}
+    loaded_confs_sources = []
+    while to_load_conf_names:
+        to_load = to_load_conf_names.pop(0)
```

Comment:
AFAIK this will never have more than one item. No need to use a collection.

Reply:
so here's the kicker:
if `~/.config/A specifies config_filename` to be 'B' instead, `/etc/xdg/A` shouldn't be read :D

Reply:
I think that is a bit too much, I'm not saying it can not be done, but I'm saying that it makes the semantic of the config chain too complicated and basically impossible to explain. If it was up to me I would keep this new semantic that is still very complicated, but it is compatible with the old one in the simple case (just one `config_filename` level). Do you agree? Else I can change it to be as you described